src/tooltip.tsx

```diff
@@ -135,10 +142,64 @@ const Tooltip: React.FC<Props> = ({
         height: 0,
     });
 
-    const getPosition = (position: Position = defaultPositionDesktop) =>
-        isTabletOrSmaller && (position === 'left' || position === 'right') ? defaultPositionMobile : position;
+    const [containerPosition, setContainerPosition] = React.useState({});
+
+    const getPosition = (position: Position | undefined) =>
+        isTabletOrSmaller ? position || defaultPositionMobile : position || defaultPositionDesktop;
+
+    const noRightSpace = (position: Position) => {
+        if (!tooltipRef.current) return position;
+        const tooltipBoundingClientRect = tooltipRef.current.getBoundingClientRect();
+
+        return targetBoundingClientRect.current.right + tooltipBoundingClientRect.width > window.innerWidth;
+    };
+
+    const noLeftSpace = (position: Position) => {
+        if (!tooltipRef.current) return position;
+        const tooltipBoundingClientRect = tooltipRef.current.getBoundingClientRect();
+
+        return targetBoundingClientRect.current.left < tooltipBoundingClientRect.width;
+    };
+
+    const validatePosition = (position: Position) => {
+        if (!tooltipRef.current) return position;
+        const tooltipBoundingClientRect = tooltipRef.current.getBoundingClientRect();
+
+        const validateLeft = (position: Position) => {
+            const noTopSpace =
+                targetBoundingClientRect.current.top < tooltipBoundingClientRect.height ? 'bottom' : 'top';
+            return noLeftSpace(position) ? noTopSpace : position;
+        };
+
+        const validateRight = (position: Position) => {
+            return noRightSpace(position) ? validateLeft('left') : position;
+        };
+
+        const positionValidated = {
+            top:
+                targetBoundingClientRect.current.top < tooltipBoundingClientRect.height
+                    ? validateRight('bottom')
+                    : validateRight(position),
+            right: validateRight(position),
+            left:
+                targetBoundingClientRect.current.left < tooltipBoundingClientRect.width
+                    ? noRightSpace(position)
+                        ? targetBoundingClientRect.current.top < tooltipBoundingClientRect.height
+                            ? 'bottom'
+                            : 'top'
+                        : 'right'
+                    : position,
+            bottom:
+                targetBoundingClientRect.current.bottom + tooltipBoundingClientRect.height >
+                window.innerHeight
+                    ? validateRight('top')
+                    : validateRight(position),
+        };
+
+        return positionValidated[position];
+    };
 
-    const position = getPosition(rest.position);
+    const position = validatePosition(getPosition(rest.position));
```

Comment:
your are making dom calculation `getBoundingClientRect` inisde `validatePosition`, you can't do that at render time, you should do that inside a react effect